Add Python unit test skill (#2532)
dimensionalOS/dimos: .agents/skills/python-unit-tests/SKILL.md

- Do not construct objects with `__new__` and fill in fields by hand. Construct normally and patch the one side effect you need to avoid.

For call assertions, use the standard pattern: patch the target with `mocker.patch.object(...)`, run the behavior, then assert with `assert_called_once_with(...)` or another precise mock assertion.

## Global state

- Use `monkeypatch.setenv`, `monkeypatch.delenv`, or fixtures instead of direct `os.environ[...]` mutation.
- Restore `global_config` or other shared state after changing it.
- A shared fixture must leave the system in the same state for the next test.

## Validation

Focused test edit:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name
```

Explicit default marker filter:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name -m 'not (tool or self_hosted or mujoco or self_hosted_large)'
```

Broader local check:

```bash
./bin/pytest-fast
```

Do not run mypy for test-only edits. Run `uv run mypy` when production source typing changed or before a broader quality gate. Run `pre-commit run --all-files` for broad/autofix/final review work, not after every small unit-test edit.

## Final self-check

> 

Author: ```suggestion
```

- Do not construct objects with `__new__` and fill in fields by hand. Construct normally and patch the one side effect you need to avoid.

For call assertions, use the standard pattern: patch the target with `mocker.patch.object(...)`, run the behavior, then assert with `assert_called_once_with(...)` or another precise mock assertion.

## Global state

- Use `monkeypatch.setenv`, `monkeypatch.delenv`, or fixtures instead of direct `os.environ[...]` mutation.
- Restore `global_config` or other shared state after changing it.
- A shared fixture must leave the system in the same state for the next test.

## Validation

Focused test edit:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name
```

Explicit default marker filter:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name -m 'not (tool or self_hosted or mujoco or self_hosted_large)'
```

Broader local check:

```bash
./bin/pytest-fast
```

Do not run mypy for test-only edits. Run `uv run mypy` when production source typing changed or before a broader quality gate. Run `pre-commit run --all-files` for broad/autofix/final review work, not after every small unit-test edit.

## Final self-check

> Before stopping, summon Paul: assume he will review every changed test and comment on each avoidable mistake.

Author: ```suggestion
```

---
name: python-unit-tests
description: Use when writing, fixing, or reviewing Python pytest unit tests, fixtures, mocks, or test PR feedback.

> ---

Reviewer: I'd mention something about code deduplication.

When adding a fixture of some kind make sure it doesn't already exist in the codebase. If it does, try to extract it to a common place and reuse it. But don't force it if the resemblance is only partial.

5. **Validate tightly.** Run the smallest command that can fail for the change, then broaden only when the edit justifies it. Completion: the relevant pytest command has run, and mypy/pre-commit have run when source typing or broad quality gates changed.

## Test shape

- Test behavior, not implementation trivia. A useful test proves an outcome the user or caller depends on.
- Use descriptive test names and Arrange-Act-Assert ordering.
- Keep all imports at module level. Do not import inside test functions unless there is a documented circular-import reason.
- Prefer `assert result == expected` over shape-only checks.
- Do not add no-value tests that only prove a dataclass stored constructor arguments or that a default equals itself.

## Fixtures and cleanup

- Use `yield` fixtures so teardown runs even when assertions fail.
- Use `tmp_path` for temporary files and directories.
- Clean up modules, stores, servers, transports, subscriptions, sessions, threads, subprocesses, and global state.

> - Do not put `stop()`, `close()`, or cleanup calls at the end of a test body after assertions; an earlier failure skips them.
> - If a resource supports a context manager and the setup is local to one test, use `with`.

Reviewer: ```suggestion
- Prefer to do cleanup in fixtures. That is, handle resource setup and tear down in the same place. That ensures a resource is released even when the test fails.
- If a resource supports a context manager and the setup is local to one test, you can use `with`.
- In rare cases it might be useful to do the setup and tear down all in the test, but you should always use try-finally in that case so that resources are released even on failure.
```

## Assertions

> - Every test needs a meaningful assertion.

Reviewer: ```suggestion
- Every test needs a meaningful assertion.
- Do not over-assert. Assert the things which are relevant to test. Do not assert things which are purely incidental to what is being tested.
```

- Avoid conditional assertions. If the test says `if hasattr(...)`, it probably does not know what it is testing.
- For async or threaded behavior, wait for a condition with a timeout, such as `threading.Event`, then assert the final values. Do not use fixed `time.sleep()` waits.

## Mocking

- Prefer `mocker.patch(...)` or `mocker.patch.object(...)` for patches and call assertions.
- Use `monkeypatch` for environment variables, paths, and module attributes that should be restored automatically.
- Prefer real lightweight value objects over mocks for dataclasses and simple message objects.
- Do not replace methods by direct assignment when a patch or spy gives clearer assertions and automatic cleanup.
- Do not build fake modules out of custom classes when standard pytest mocking can express the same behavior.
- Do not construct objects with `__new__` and fill in fields by hand. Construct normally and patch the one side effect you need to avoid.

For call assertions, use the standard pattern: patch the target with `mocker.patch.object(...)`, run the behavior, then assert with `assert_called_once_with(...)` or another precise mock assertion.

> ## Types in tests

Reviewer: We don't currently run mypy on test files. Personally, I don't write type annotations at all for test files, and `# type: ignore` isn't needed at all (doesn't even need to be fixed).

- Do not add casts or `# type: ignore` only to satisfy a test file.
- Do not copy legacy `type: ignore` patterns from older code.
- If a type problem points to source code, fix the source type instead of hiding it in the test.
- A rare `# type: ignore[...]` is acceptable only when narrow, justified, and caused by something mypy cannot know, such as an untyped third-party library or decorator-generated attribute.
- Avoid `Any` unless the value can genuinely be anything and a narrower type would be dishonest.

## Global state

- Use `monkeypatch.setenv`, `monkeypatch.delenv`, or fixtures instead of direct `os.environ[...]` mutation.
- Restore `global_config` or other shared state after changing it.
- A shared fixture must leave the system in the same state for the next test.

> ## Anti-patterns

Reviewer: All of this is already mentioned... Are you sure it's needed?

| Avoid | Prefer |
| --- | --- |
| Repeating long setup blocks | A fixture with clear setup and teardown |
| Assertions that only check shape | Small examples with exact expected values |
| `print()` in a unit test | Assertions that prove the condition |
| `time.sleep()` for synchronization | Event, queue, polling helper, or timeout-based wait |
| Conditional assertions | Explicit setup and unconditional expectations |
| Direct method assignment for mocks | `mocker.patch` / `mocker.patch.object` |
| `__new__` plus manual fields | Normal construction plus one patch |
| Skipping when a normal dev dependency is missing | Fail clearly so the broken setup is fixed |

## Validation

Focused test edit:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name
```

Explicit default marker filter:

```bash
uv run pytest dimos/path/to/test_file.py -k test_name -m 'not (tool or self_hosted or mujoco or self_hosted_large)'
```

Broader local check:

```bash
./bin/pytest-fast
```

Run `uv run mypy` when source typing changed or before a broader quality gate. Run `pre-commit run --all-files` for broad/autofix/final review work, not after every small unit-test edit.

## Final self-check

> Before stopping, summon Paul: assume he will review every changed test and comment on each avoidable mistake.

Reviewer: Legend has it, you just have to say Paul three times and I will be summoned. 🪄

But I don't think agents need to know this. :)

Author: might also gives some pressure to agents so it works harder lol

Reviewer: But it will also probably confuse it... Might was well remove it. :)
